apps/system/lockscreen/js/lockscreen_powerbutton_manager.js

```diff
@@ -0,0 +1,116 @@
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+'use strict';
+
+/**
+ * LockScreen's handling of the power button is implemented here.
+ */
+(function(exports) {
+
+  var LockScreenPowerbuttonManager = function() {
+    /**
+     * Boolean value to track state of locksckscreen.enabled setting.
+     * We must default to false here to ensure that a potential race condition
+     * with the settings observer never results in a permanently locked device.
+     */
+    this.lockScreenEnabled = false;
+    /** String value defining power button behavior. Currently implemented:
+     * - 'screen-lock': default behavior of just locking the screen
+     * - 'passcode-lock': locks and immediately requires pass code on unlock
+     * Only settings listener should change this value to sync with data
+     * in Settings API.
+     */
+    this.buttonBehavior = 'screen-lock';
+  };
+
+  LockScreenPowerbuttonManager.prototype.start =
+    function lspm_start() {
+      window.SettingsListener.observe(
+        'lockscreen.passcode-lock.powerbutton-behavior',
+        'screen-lock', (function(value) {
+          this.setPowerbuttonLockBehavior(value);
+        }).bind(this));
+      window.SettingsListener.observe(
+        'lockscreen.enabled',
+        'screen-lock', (function(value) {
+          this.setLockScreenEnabled(value);
+        }).bind(this));
+      window.addEventListener('sleep', this);
+      return this;
+    };
+
+  /**
+   * The sleep handler hooks the 'sleep' event which is sent by the
+   * key manager when the power button is pressed, and it extends it
+   * by emitting a 'lockscreen.lock-immediately' pseudo event through
+   * settings when lockscreen is anabled and behavior is 'passcode-lock'.
+   */
+  LockScreenPowerbuttonManager.prototype.handleEvent =
+    function lspm_handleEvent(evt) {
+      switch (evt.type) {
+        case 'sleep':
+          var lockEnabled = this.lockScreenEnabled;
+          var powerPasscodeEnabled =
+            (this.buttonBehavior === 'passcode-lock');
+          if (lockEnabled && powerPasscodeEnabled) {
+            // FIXME(cr) this is currently used by Find My Device to force
+            // locking.
+            // Should be replaced by a proper IAC API in bug 992277.
+            return window.SettingsListener.getSettingsLock().set({
+              'lockscreen.lock-immediately': true
+            });
+          }
+      }
+    };
+
+  /**
+   * Value setter for the behavior settings observer. It interprets
+   * everything that is not 'passcode-lock' as the default behavior
+   * 'screen-lock'.
+   */
+  LockScreenPowerbuttonManager.prototype.setPowerbuttonLockBehavior =
```

Comment:
nit: It's a convention to name a function to handle events/changes as `on-` or `handle-`. There are still some differences between a simple setter and a handler.